From: Mark Tinguely <tinguely@sgi.com>
Subject: xfs_dmapi: fix value from newer Linux strnlen_user()
References: bsc#932897
Patch-mainline: Never (dmapi isn't upstream)

Community commit v3.4-7800-ga08c535 implemented a generic
strnlen_user(). That function can return the length of the
user string but fail to restrict it to the supplied maximum.
This bug will cause an overwrite of the DMAPI name and
triggers the stack corruption detection.

Manually restrict the string length by comparing the
strnlen_user() result to our supplied maximum.

Signed-off-by: Mark Tinguely <tinguely@sgi.com>
Reviewed-by: Olaf Weber <olaf@sgi.com>
Acked-by: Jan Kara <jack@suse.cz>

---
 fs/xfs/dmapi/xfs_dm.c |    7 +++++++
 1 file changed, 7 insertions(+)

--- a/fs/xfs/dmapi/xfs_dm.c
+++ b/fs/xfs/dmapi/xfs_dm.c
@@ -249,6 +249,13 @@ xfs_copyin_attrname(
 	strcpy(to->dan_chars, dmattr_prefix);
 
         len = strnlen_user((char __user *)from, DM_ATTR_NAME_SIZE);
+	/*
+	 * strnlen_user() is broken on newer version of Linux.
+	 * fix by manually setting the upper limit.
+	 * strnlen_user() will also count the NULL.
+	 */
+	if (len > DM_ATTR_NAME_SIZE + 1)
+		len = DM_ATTR_NAME_SIZE + 1;
         if (len == 0)
             error = EFAULT;
         else {
